src/connection.rs

```diff
@@ -514,12 +402,66 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Connection<T> {
 
                     return Err(ConnectionError::Closed);
                 }
-                Either::Right((frame, _)) => {
-                    if let Some(stream) = self.on_frame(frame.transpose().map_err(Into::into))? {
-                        return Ok(stream);
+                Event::NewStream(stream) => {
+                    return Ok(stream);
+                }
+            }
+        }
+    }
+
+    fn poll(&mut self, cx: &mut Context<'_>) -> Poll<Result<Event>> {
+        loop {
+            self.garbage_collect();
+
+            if self.socket.poll_ready_unpin(cx).is_ready() {
+                if let Some(frame) = self.pending_frames.pop_front() {
+                    self.socket.start_send_unpin(frame)?;
+                    continue;
+                }
+            }
+
+            match self.socket.poll_flush_unpin(cx)? {
+                Poll::Ready(()) => {}
+                Poll::Pending => {}
+            }
+
+            match self.socket.poll_next_unpin(cx) {
+                Poll::Ready(Some(frame)) => {
+                    if let Some(stream) = self.on_frame(frame?)? {
+                        return Poll::Ready(Ok(Event::NewStream(stream)));
                     }
+                    continue;
                 }
+                Poll::Ready(None) => {
+                    return Poll::Ready(Err(ConnectionError::Closed));
+                }
+                _ => {}
+            }
+
+            match self.control_receiver.poll_next_unpin(cx) {
+                Poll::Ready(Some(cmd)) => {
+                    self.on_control_command(cmd)?;
+                    continue;
+                }
+                Poll::Ready(None) => {
+                    return Poll::Ready(Ok(Event::ControlReceiverClosed));
+                }
+                _ => {}
```

Comment:
Why not write out `Poll::Pending` here? Isn't it the only other variant? In my eyes would be more descriptive. Same for the below.